File: June-LeetCoding-Challenge-2020/Week3/174-dungeon-game.py
class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        height, width = len(dungeon), len(dungeon[0])
        m,n = height-1, width-1
        
        dp = dungeon
        # where the princess is
        dp[m][n] = max(1,1-dp[m][n])
        
        for r in range(m-1, -1, -1):
            dp[r][n] = max(1,dp[r+1][n]-dp[r][n])
            
        for c in range(n-1, -1, -1):
            dp[m][c] = max(1, dp[m][c+1]-dp[m][c])
            
        for r in range(m-1, -1, -1):
            for c in range(n-1, -1, -1):
                dp[r][c] = max(1,min(dp[r+1][c], dp[r][c+1])-dp[r][c])
                
        return dp[0][0]
    # TC: O(N^2)
    # SC: O(N^2)

File: June-LeetCoding-Challenge-2020/Week3/174-dungeon-game.cpp
// Solution #1 Botton Up Dynamic Programming
// TC: O(row*col size) has to traverse row*col
// SC: O(size of column) storing value
// Runtime: 8 ms, faster than 96.89%
// Memory Usage: 9.1 MB, less than 34.39%
class Solution {
public:
    
    int calculateMinimumHP(vector<vector<int> > &dungeon) {
        const int m = dungeon.size();
        const int n = dungeon[0].size();
        
        vector<int> dp(n + 1, INT_MAX);
        
        dp[n - 1] = 0; 
        
        for(int i = m - 1; i >= 0; i--)
            for(int j = n - 1; j >= 0; j--)
                dp[j] = min(dp[j], dp[j + 1]) - dungeon[i][j] >= 0 ? min(dp[j], dp[j + 1]) - dungeon[i][j] : 0;
        

        return dp[0] + 1;
    }

};